File: sdks/dart/lib/src/server/billing.dart
import '../server_client.dart';

class Billing {
  final ApplAdServer _client;

  Billing(this._client);

  Future<Map<String, dynamic>> listPlans() async {
    return _client.get('/v1/billing/plans');
  }

  Future<Map<String, dynamic>> getSubscription() async {
    return _client.get('/v1/billing/subscription');
  }

  Future<Map<String, dynamic>> subscribe({
    required String planId,
    String? paymentMethodId,
  }) async {
    return _client.post('/v1/billing/subscription', data: {
      'planId': planId,
      if (paymentMethodId != null) 'paymentMethodId': paymentMethodId,
    });
  }

  Future<Map<String, dynamic>> cancelSubscription() async {
    return _client.delete('/v1/billing/subscription');
  }

  Future<Map<String, dynamic>> getUsage() async {
    return _client.get('/v1/billing/usage');
  }

  Future<Map<String, dynamic>> listInvoices() async {
    return _client.get('/v1/billing/invoices');
  }

  Future<Map<String, dynamic>> getInvoice(String invoiceId) async {
    return _client.get('/v1/billing/invoices/$invoiceId');
  }
}

File: sdks/python/applad/billing.py
"""Billing service."""


class Billing:
    def __init__(self, client):
        self.client = client

    def list_plans(self):
        return self.client._call("GET", "/billing/plans")

    def get_subscription(self):
        return self.client._call("GET", "/billing/subscription")

    def subscribe(self, plan_id: str, payment_method_id: str = None):
        data = {"planId": plan_id}
        if payment_method_id is not None:
            data["paymentMethodId"] = payment_method_id
        return self.client._call("POST", "/billing/subscription", data)

    def cancel_subscription(self):
        return self.client._call("DELETE", "/billing/subscription")

    def get_usage(self):
        return self.client._call("GET", "/billing/usage")

    def list_invoices(self):
        return self.client._call("GET", "/billing/invoices")

    def get_invoice(self, invoice_id: str):
        return self.client._call("GET", f"/billing/invoices/{invoice_id}")